[FLINK-15020][hive] Support timestamp type in hive

lirui-apache opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Support read/write timestamp in Hive


## Brief change log

  - Map timestamp type in Hive to `DataTypes.TIMESTAMP(9)`. Since Hive doesn't support timestamp precision, only timestamps of precision 9 is supported by HiveCatalog at the moment.
  - Add shim method to convert timestamp instances between Hive and Flink.
  - New test case for read/write timestamp.


## Verifying this change

New test case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

lirui-apache commented:
cc @JingsongLi @bowenli86 @xuefuz @zjuwangg

bowenli86 commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveReflectionUtils.java:
```diff
@@ -82,12 +86,34 @@ public static Object convertToHiveDate(HiveShim hiveShim, String s) throws Flink
+	// converts a Flink timestamp instance to what's expected by Hive
+	public static Object toHiveTimestamp(HiveShim hiveShim, Object flinkTimestamp) {
+		Preconditions.checkArgument(flinkTimestamp instanceof Timestamp || flinkTimestamp instanceof LocalDateTime,
+				String.format("Only support converting %s or %s to Hive timestamp, but got %s",
```
```suggestion
				String.format("Only support converting %s or %s to Hive timestamp, but not %s",
```

JingsongLi replied:
Please not use `String.format` in `checkArgument`, this is a per record column operation. And `String.format` is very expensive, we should avoid this. Please use:
```
public static void checkArgument(boolean condition,
			@Nullable String errorMessageTemplate,
			@Nullable Object... errorMessageArgs) {
    ......
}
```

bowenli86 commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java:
```diff
@@ -244,7 +247,11 @@ public static Object toFlinkObject(ObjectInspector inspector, Object data) {
 				PrimitiveObjectInspector poi = (PrimitiveObjectInspector) inspector;
-				return poi.getPrimitiveJavaObject(data);
+				Object javaObject = poi.getPrimitiveJavaObject(data);
+				if (inspector instanceof TimestampObjectInspector) {
```
shall we move `inspector instanceof TimestampObjectInspector` to its own else-if? rather than doing it twice here

JingsongLi replied:
+1

bowenli86 commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java:
```diff
@@ -436,6 +440,7 @@ public static DataType toFlinkType(ObjectInspector inspector) {
 					case DATE:
 						return Date.class;
 					case TIMESTAMP:
+						return LocalDateTime.class;
```
why is this LocalDateTime.class rather than Timestamp.class?

lirui-apache replied:
Flink timestamp by default bridges to LocalDateTime

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java:
```diff
@@ -274,6 +274,9 @@ public TypeInfo visit(DateType dateType) {
 
 		@Override
 		public TypeInfo visit(TimestampType timestampType) {
+			if (timestampType.getPrecision() != 9) {
```
I'm wondering if this is too restrictive. While Hive only supports a precision of 9, it should be able to handle any ts type with a precision less than 9.  Thus, Hive timestamp needs to be mapped to Flink ts(9) while any ts(<=9) from Flink can be mapped to Hive timestamp. What do you think?

lirui-apache replied:
Ideally yes, we should support any ts(<=9). But to do that, we need a way to store the precision of timestamp in HMS, and restore the precision when retrieving the table schema. I prefer to leave it to follow up tasks.

JingsongLi replied:
-1 for this restrictive check.
It is very important that we use this way to support UDF too. If we have this restrictive check here, UDF use timestamp type will be very restrictive.
I prefer to revolve this problem now.

lirui-apache replied:
The restriction is maintained when creating table in HiveCatalog. Added a switch to turn off the check so that functions calling this util method can support different precisions.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveReflectionUtils.java:
```diff
@@ -82,12 +86,34 @@ public static Object convertToHiveDate(HiveShim hiveShim, String s) throws Flink
+		if (hiveTimestamp instanceof Timestamp) {
+			return (Timestamp) hiveTimestamp;
+		}
 		try {
```
We should probably put an assertion on the class name so that we will know if things break.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveReflectionUtils.java:
```diff
@@ -82,12 +86,34 @@ public static Object convertToHiveDate(HiveShim hiveShim, String s) throws Flink
-			throw new FlinkHiveUDFException("Failed to invoke Hive's Timestamp.valueOf()", e);
+			String hiveTSStr = (String) invokeMethod(hiveShim.getTimestampDataTypeClass(), hiveTimestamp,
+					"toString", null, null);
+			return Timestamp.valueOf(hiveTSStr);
```
The conversion to/from string might be two slow for the conversion. We probably can get milli seconds and nano seconds from java.sql.timestamp and set them to Hive's timestamp.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveReflectionUtils.java:
```diff
@@ -82,12 +86,34 @@ public static Object convertToHiveDate(HiveShim hiveShim, String s) throws Flink
+				return invokeMethod(hiveTimestampClz, null, "valueOf", new Class[]{String.class}, new Object[]{flinkTimestamp.toString()});
+			} catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+				throw new FlinkHiveException("Failed to convert to Hive timestamp", e);
+			}
```
1. An assertion on class name might help.
2. Conversion to/from string might be slow. We probably can set milli and nano seconds directly.

lirui-apache replied:
For #1, maybe it's better to add a shim method for the conversion, rather than assume class names in `HiveReflectionUtils`.

docete commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java:
```diff
@@ -436,6 +440,7 @@ public static DataType toFlinkType(ObjectInspector inspector) {
 					case DATE:
 						return Date.class;
 					case TIMESTAMP:
+						return LocalDateTime.class;
```
internally flink use the new SqlTimestamp to represent a java.sql.Timestmap/java.time.LocalDateTime object. and by default, TIMESTAMP type would convert to java.time.LocalDateTime when leaving the table ecosystem.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveReflectionUtils.java:
```diff
@@ -82,12 +86,34 @@ public static Object convertToHiveDate(HiveShim hiveShim, String s) throws Flink
+			return flinkTimestamp instanceof Timestamp ? flinkTimestamp : Timestamp.valueOf((LocalDateTime) flinkTimestamp);
+		} else {
+			try {
+				return invokeMethod(hiveTimestampClz, null, "valueOf", new Class[]{String.class}, new Object[]{flinkTimestamp.toString()});
```
This is a per record column operation. And java reflection is very expensive. You should extract this method to reuse.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveReflectionUtils.java:
```diff
@@ -82,12 +86,34 @@ public static Object convertToHiveDate(HiveShim hiveShim, String s) throws Flink
-			return method.invoke(null, s);
-		} catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
-			throw new FlinkHiveUDFException("Failed to invoke Hive's Timestamp.valueOf()", e);
+			String hiveTSStr = (String) invokeMethod(hiveShim.getTimestampDataTypeClass(), hiveTimestamp,
```
This is a per record column operation. And java reflection is very expensive. You should extract this method to reuse.

JingsongLi commented on flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorTest.java:
```diff
@@ -328,6 +330,31 @@ public void testNotNullConstraints() throws Exception {
 	}
 
+	@Test
+	public void testTimestamp() throws Exception {
```
Please add tests to `testDifferentFormats` too, because some formats like orc are total another way.

lirui-apache replied:
Added test in `testDifferentFormats ` but there's some issue with the vectorized ORC reader, so I temporarily turned off the reader. Please have a look.

JingsongLi replied:
Please not turn off whole reader. You can modify `HiveTableInputFormat.useOrcVectorizedRead`.

JingsongLi commented on flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java:
```diff
@@ -133,7 +134,7 @@ public void testUDFWeekOfYear() throws FlinkHiveUDFException {
 		assertEquals(29, udf.eval("1969-07-20"));
 		assertEquals(29, udf.eval(HiveReflectionUtils.convertToHiveDate(hiveShim, "1969-07-20")));
-		assertEquals(29, udf.eval(HiveReflectionUtils.convertToHiveTimestamp(hiveShim, "1969-07-20 00:00:00")));
+		assertEquals(29, udf.eval(HiveReflectionUtils.toHiveTimestamp(hiveShim, Timestamp.valueOf("1969-07-20 00:00:00"))));
```
Please add tests to `HiveCatalogUseBlinkITCase` too. Because that is it case. And now there is no it case to cover UDX.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java:
```diff
@@ -401,10 +409,6 @@ private static ObjectInspector getObjectInspector(TypeInfo type) {
-	}
-
 	// given a Hive ObjectInspector, get the class for corresponding Flink object
 	private static Class<?> getClassFromObjectInspector(ObjectInspector inspector) {
```
My suggestion is just use `toFlinkType(inspector).getConversionClass`.

lirui-apache commented:
@xuefuz @JingsongLi @docete Thanks for your review. PR updated to:

1.  Move timestamp conversion between Hive and Flink into HiveShim
2. Avoid parsing timestamp strings
3. Reuse reflection class, constructor, etc.
4. Lift the timestamp precision restriction for UDFs
5. Add more test cases
6. Address other comments

Please have another look.

lirui-apache commented:
Also ping @bowenli86 for further review. Thanks.

lirui-apache commented:
@xuefuz I just refactored the code such that callers need to explicitly specify whether precision check is needed. Hopefully that'll make things clearer. The conversion from `DataType` to `TypeInfo` is needed in the following code paths:

1. When creating tables via HiveCatalog. As we discussed earlier, we'll enforce the check in this case.
2. When getting `ObjectInspector` for writing to Hive tables. Since Hive tables only support ts(9), we'll enforce the check here.
3. Used by Hive UDFs during initialization, deciding result type, getting `ObjectInspector`s for parameters, etc. We don't enforce the check in this case since we want UDFs to be able to handle ts (<9).

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableInputFormat.java:
```diff
@@ -170,8 +172,9 @@ private boolean useOrcVectorizedRead(HiveTablePartition partition) {
 
-		LOG.info("Use flink orc ColumnarRow reader.");
-		return true;
+		// temporarily disable the vectorized reader, until we figure out how to read timestamps with it
```
No... I mean modify timestamp type, you can see above, there are support types.
I think you can use https://github.com/apache/flink/pull/10426 to verify and wait a minute to merge.

xuefuz commented:
> @xuefuz I just refactored the code such that callers need to explicitly specify whether precision check is needed. Hopefully that'll make things clearer. The conversion from `DataType` to `TypeInfo` is needed in the following code paths:
> 
>     1. When creating tables via HiveCatalog. As we discussed earlier, we'll enforce the check in this case.
> 
>     2. When getting `ObjectInspector` for writing to Hive tables. Since Hive tables only support ts(9), we'll enforce the check here.
> 
>     3. Used by Hive UDFs during initialization, deciding result type, getting `ObjectInspector`s for parameters, etc. We don't enforce the check in this case since we want UDFs to be able to handle ts (<9).

Thanks for the explanation. It sounds good. One thing to watch out is when Flink timestamp value has greater precision than 9, which can lose precision after converting to Hive timestamp. We don't have to address this in the PR, but it's one interesting thing to find out.

I'm +1 for this PR to merge. For orc reader issue. if #10426 fix the issue, we can merge either one first, and the second to turn the switch back on.

lirui-apache commented:
@JingsongLi Got your point. Just updated to just disable the vector reader for timestamp.
Latest build passed for my personal repo: https://travis-ci.org/lirui-apache/flink/builds/620936056?utm_medium=notification&utm_source=email

JingsongLi commented:
https://github.com/apache/flink/pull/10426 merged, I think you can rebase master and open timestamp type for new orc reader now.

lirui-apache commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSimpleUDF.java:
```diff
@@ -57,7 +57,7 @@
 	private transient HiveObjectConversion[] conversions;
 	private transient boolean allIdentityConverter;
-	private transient HiveShim hiveShim;
+	private HiveShim hiveShim;
```
Because it's needed after deserialization to call the util methods

lirui-apache commented:
Rebased and re-enabled the vector reader for timestamp.
@JingsongLi @bowenli86 Please let me know if you have further comments, thanks.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV310.java:
```diff
@@ -43,6 +50,30 @@
 public class HiveShimV310 extends HiveShimV235 {
 
+	private transient Class hiveTimestampClz;
+	private transient Constructor hiveTimestampConstructor;
```
Can we just use static?

JingsongLi commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java:
```diff
@@ -252,6 +253,11 @@ public static DataFormatConverter getConverterForDataType(DataType originDataTyp
 					return BinaryGenericConverter.INSTANCE;
 				}
 				return new GenericConverter(typeInfo.createSerializer(new ExecutionConfig()));
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
```
https://github.com/apache/flink/pull/10418

lirui-apache replied:
Do you mean we should wait for #10418 to get in first?

JingsongLi replied:
No, you can discuss with @docete

bowenli86 commented:
LGTM. can you open a ticket for updating documentation?

lirui-apache commented:
Rebased and addressed comments.
I think we can merge either this one or #10418 first, and the second one just rebase and resolve conflicts.
cc @JingsongLi @docete @KurtYoung

lirui-apache commented:
Passed for my own repo: https://travis-ci.org/lirui-apache/flink/builds/621925513

bowenli86 commented:
LGTM, merging

lirui-apache commented:
> LGTM. can you open a ticket for updating documentation?

Opened FLINK-15128 for it.
